python/core/qgsprojectstorage.sip.in

```diff
@@ -0,0 +1,115 @@
+/************************************************************************
+ * This file has been generated automatically from                      *
+ *                                                                      *
+ * src/core/qgsprojectstorage.h                                         *
+ *                                                                      *
+ * Do not edit manually ! Edit header and run scripts/sipify.pl again   *
+ ************************************************************************/
+
+
+
+
+
+
+class QgsProjectStorage
+{
+%Docstring
+Abstract interface for project storage - to be implemented by various backends
+and registered in QgsProjectStorageRegistry.
+
+.. versionadded:: 3.2
+%End
+
+%TypeHeaderCode
+#include "qgsprojectstorage.h"
+%End
+  public:
+
+    class Metadata
+{
+%Docstring
+Metadata associated with a project
+%End
+
+%TypeHeaderCode
+#include "qgsprojectstorage.h"
+%End
+      public:
+        QString name;
+        QDateTime lastModified;
```

Comment:
I think we should use a variant for lastModified checks and storage - it's possible some providers won't have access to a last modified time and may instead need to use a version number or project hash instead.

Reply:
I would really prefer to stick with QDateTime - if there would be a storage implementation that can't support last modified time, I would prefer to introduce some other member(s) to accommodate more kinds of metadata (or just a QVariantMap for any extra stuff). Some implementations would keep the time value invalid in the worst case... Would that work for you? I am trying to avoid variants as much as possible as they suddenly make all the nice c++ type checks disappear...

Reply:
I'd be happy to see a QVariantMap added, but that kind-of goes against your desire to avoid variants. I guess in either case we need a `bool QgsProjectStorage::hasProjectChanged( const QgsProjectStorage::Metadata& metadata )` member in QgsProjectStorage, right?

Reply:
I think we would need a hasProjectChanged() virtual method only if we were to introduce support for project storage implementation that cannot store last modified time for some reason. For now QGIS app just compares timestamps - in the future when needed the extension to use hasProjectChanged() should be fairly straightforward...